### PR TITLE
Extending support for mixing binders and terms in abbreviations

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** feature

By consistency with notations modifying the parser, add support for abbreviations which bind both a term and a pattern:
```
Notation U t := (t,fun t => t).
Check U tt.
(* gives "(tt, fun 'tt => tt)", printed "U tt" *)
Check fun x => U x.
(* gives "fun x : ?A => (x, fun x0 : ?T => x0)", printed "fun x : ?A => U x" *)
```

This is virtually a source of incompatibilities if a notation was already using a binder with same name as a notation variable.

Synchronous overlay: LPCIC/coq-elpi#130

- [X] Added / updated test-suite

### CODE REVIEW

**Comment by mattam82** [comment]
That looks quite strange to me, what is the use case? If I declare the x as
a pattern but use it as an arbitrary term in the notation body this is
allowed?
Le mar. 23 oct. 2018 à 17:00, Hugo Herbelin <notifications@github.com> a
écrit :

**Comment by herbelin** [comment]
@mattam82, here is a typical example from Iris of notations _modifying the parser_ and using variables in both term and binder position:
```coq
Inductive ftele : Type :=
| fb {T:Type} : T -> ftele
| fr {T} : (T -> ftele) -> ftele.
Fixpoint args ftele : Type :=
  match ftele with
    | fb _ => unit
    | fr f => sigT (fun t => args (f t))
  end.
Definition fpack := sigT args.
Definition pack fp fa : fpack := existT _ fp fa.
Notation "'tele' x .. z := b" :=
  (fun x => .. (fun z =>
     pack (fr (fun x => .. ( fr (fun z => fb b) ) .. ) )
          (existT _ x .. (existT _ z tt) .. )
                ) ..)
  (at level 85, x binder, z binder).
```

@ejgallego
> I was under the impression that syndefs were frozen and to be replaced by notations. Isn't that the case?

Syndefs (i.e. abbreviations, i.e. notations of the form `Notation f args := ...`) are quite popular, especially because they don't modify the parser. I don't see how we could renounce to them.

**Comment by herbelin** [comment]
Any other opinion on whether it is more useful or more troubling to allow abbreviations to bind variables as in the example above?

One could be more restrictive and that would be easy to implement (maybe no user is anyway really using the ability to bind variables in abbreviations). Or one could continue to support them (that would also be easy).

Whatever decision is taken, this PR is also doing some code factorization which is worth to be integrated. So, in the absence of strong opinion, the status quo would imply to merge this cleanup as it. But, @mattam82, if you wish, I can also remove the ability to write the strange things above.

**Comment by herbelin** [comment]
@mattam82: I don't think this need a lot of thinking time. It is either about the status quo and merge the PR which improves over the status quo, or being stricter about binders in abbreviations and I can also implement this.

**Comment by mattam82** [comment]
@herbelin Thanks for the explanations. I'm convinced by the Iris example that you want that feature for Notations, so I agree with your assessment, I'll merge when CI passes.

**Review by mattam82 — APPROVED** [review]

File: vernac/metasyntax.ml
Comment (mattam82) [inline vernac/metasyntax.ml]: This new warning should be documented.

**Review by mattam82 — DISMISSED** [review]

File: vernac/metasyntax.ml
Comment (herbelin) [reply vernac/metasyntax.ml]: Then I need help from @zimmi48 - I don't know where this documentation should go.

**Comment by herbelin** [comment]
> I'm convinced by the Iris example that you want that feature for Notations

Note that Iris uses parsing-level notations for that (afaik), not abbreviations. In particular, it is unclear that use of binders in abbreviations (nametab-level notations) is actually common in the large (maybe in MathComp?).

File: vernac/metasyntax.ml
Comment (Zimmi48) [reply vernac/metasyntax.ml]: It looks like this should be documented in the notation chapter: https://coq.inria.fr/refman/user-extensions/syntax-extensions.html#abbreviations
You should use the `.. warn::` directive to give the name of the warning.

**Comment by mattam82** [comment]
@herbelin this needs a rebase apparently

**Comment by mattam82** [comment]
@herbelin This should be easy to merge once rebased too.

**Comment by herbelin** [comment]
I'm finally resurrecting this PR as part of an ongoing uniformization of notations and abbreviations.

I added a few words in the chapter user extensions (tell me @coq/doc-maintainers if it looks ok enough).

An overlay for elpi shall be needed.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:615]: This is equivalent to the following:

```suggestion
.. coqtop:: in rest
```

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1506]:
```suggestion
   .. coqtop:: in reset
```

**Review by Zimmi48 — COMMENTED** [review]
I am too ignorant to properly review the content of this documentation. But since you ask, I don't see what more the documentation would need to contain.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1502]: rest = reset ?? I would not have guessed!

Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Thanks for the feedback and suggestions!

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst:611]: Looks like I made a typo in my suggestion:
```suggestion
.. coqtop:: in reset
```

Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst:611]: Ah, no problem.

**Comment by herbelin** [comment]
May I sollicitate @ejgallego to be an alternative assignee for this PR?

**Comment by Zimmi48** [comment]
@herbelin Looks like this still needs a changelog entry. I also suggest squashing a bit the documentation commits.

**Comment by herbelin** [comment]
> Looks like this still needs a changelog entry. I also suggest squashing a bit the documentation commits.

Done.

Elpi overlay also rebased.

**Review by ejgallego — APPROVED** [review]
Indeed the semantics of capturing substitutions seems a bit surprising, but as pointed out by @herbelin this is not new. This PR seems OK.

**Pull request merged**
